File: examples/xor/xor.cpp
#include <OpenANN/OpenANN>
#include <OpenANN/io/Logger.h>
#include <OpenANN/io/DirectStorageDataSet.h>
#include <OpenANN/util/Random.h>
#include <Eigen/Core>
#include <iostream>

using namespace OpenANN;

int main()
{
  // Create dataset
  const int D = 2; // number of inputs
  const int F = 1; // number of outputs
  const int N = 4; // size of training set
  Eigen::MatrixXd X(N, D); // inputs
  Eigen::MatrixXd T(N, F); // desired outputs (targets)
  // Each row represents an instance
  X.row(0) << 0.0, 1.0;
  T.row(0) << 1.0;
  X.row(1) << 0.0, 0.0;
  T.row(1) << 0.0;
  X.row(2) << 1.0, 1.0;
  T.row(2) << 0.0;
  X.row(3) << 1.0, 0.0;
  T.row(3) << 1.0;
  DirectStorageDataSet dataSet(&X, &T);

  // Make the result repeatable
  RandomNumberGenerator().seed(0);

  // Create network
  Net net;
  // Add an input layer with D inputs, 1 hidden layer with 2 nodes and an
  // output layer with F outputs. Use logistic activation function in hidden
  // layer and output layer.
  makeMLNN(net, LOGISTIC, LOGISTIC, D, F, 1, 2);
  // Add training set
  net.trainingSet(dataSet);

  // Set stopping conditions
  StoppingCriteria stop;
  stop.minimalValueDifferences = 1e-10;
  // Train network, i.e. minimize sum of squared errors (SSE) with
  // Levenberg-Marquardt optimization algorithm until the stopping criteria
  // are satisfied.
  train(net, "LMA", MSE, stop);

  // Use network to predict labels of the training data
  for(int n = 0; n < N; n++)
  {
    Eigen::VectorXd y = net(dataSet.getInstance(n));
    std::cout << y << std::endl;
  }

  return 0;
}

File: examples/xor/xor.py
from openann import *
import numpy

if __name__ == "__main__":
    # Create dataset
    X = numpy.array([[0, 1], [0, 0], [1, 1], [1, 0]])
    Y = numpy.array([[1], [0], [0], [1]])
    D = X.shape[1]
    F = Y.shape[1]
    N = X.shape[0]
    dataset = Dataset(X, Y)

    # Make the result repeatable
    RandomNumberGenerator().seed(0)

    # Create network
    net = Net()
    net.input_layer(D)
    net.fully_connected_layer(3, Activation.LOGISTIC)
    net.output_layer(F, Activation.LOGISTIC)

    # Train network
    stop_dict = {"minimal_value_differences" : 1e-10}
    lma = LMA(stop_dict)
    lma.optimize(net, dataset)

    # Use network
    for n in range(N):
        y = net.predict(X[n])
        print(y)
